src/Http/Http/src/Features/RequestBodyPipeFeature.cs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO.Pipelines;
+
+namespace Microsoft.AspNetCore.Http.Features
+{
+    public class RequestBodyPipeFeature : IRequestBodyPipeFeature
+    {
+        // Lambda hoisted to static readonly field to improve inlining https://github.com/dotnet/roslyn/issues/13624
+        private readonly static Func<IFeatureCollection, IHttpRequestFeature> _nullRequestFeature = f => null;
+
+        private PipeReader _pipeReader;
+        private FeatureReferences<IHttpRequestFeature> _features;
+
+        public RequestBodyPipeFeature(IFeatureCollection features)
+        {
+            if (features == null)
+            {
+                throw new ArgumentNullException(nameof(features));
+            }
+
+            _features = new FeatureReferences<IHttpRequestFeature>(features);
+        }
+
+        private IHttpRequestFeature HttpRequestFeature =>
+            _features.Fetch(ref _features.Cache, _nullRequestFeature);
+
+        public PipeReader PipeReader
+        {
+            get
+            {
+                if (_pipeReader == null)
+                {
+                    _pipeReader = new StreamPipeReader(HttpRequestFeature.Body);
+                }
+
+                return _pipeReader;
+            }
+            set
+            {
+                _pipeReader = value;
+                if (_pipeReader == null)
+                {
+                    HttpRequestFeature.Body = null;
+                }
+                else
+                {
+                    // TODO set the Response body to adapted pipe https://github.com/aspnet/AspNetCore/issues/3971
```

Comment:
I disagree. This work is more important that handling when people replace the PipeBody with a different Pipe. It isn't a big deal that this doesn't work yet; it will be the next work item though.

Reply:
That's true, this direction isn't as critical because the other direction doesn't even exist today. That said, both directions need to work in preview2 and this is incremental progress so we shouldn't block on that.